src/lib.rs

```diff
@@ -246,86 +285,115 @@ impl Client {
-                            _ = shutdown_channel.recv() => break,
-                            conn = Client::attempt_connect(&config, shutdown_channel2) => conn,
-                        }
-                    }
-                    Some(ref mut connection) => {
-                        tokio::select! {
-                            Some(msg) = command_receiver.recv() => {
-                                Client::handle_msg(msg, connection, &config, &shutdown_channel).await
-                            }
-                            _ = shutdown_channel.recv() => {
-                                let _ = connection.end().await;
-                                break;
+                let msg = tokio::select! {
+                    msg = command_receiver.recv() => match msg {
+                        Some(msg) => msg,
+
+                        // If no messages are in the channel and the sender is closed we can bail out
+                        // The sender can only be closed after all `Client`s have been dropped and the heartbeat task has ended
+                        // The heartbeat task should never go away before this task does, so this should be unreachable
+                        None => unreachable!(),
+                    },
+
+                    _ = shutdown_request_rx.recv() => {
+                        // Since we must shutdown, nothing can be enqueued anymore. If the queue is empty we bail out.
+                        // If not, we empty it and reschedule the shutdown as we just consumed the shutdown command
+                        match command_receiver.try_recv() {
+                            Err(_) => break,
+                            Ok(msg) => {
+                                // Only failures possible are Full and Close
+                                // If queue is full someone raced with us by scheduling the shutdown (by calling `Self::close`) so we don't need to do it again
+                                // Closed should be impossible as this task owns the receiver, it will never go away from under us and we never call close on it
+                                if let Err(TrySendError::Closed(_)) = shutdown_request_tx.try_send(()) {
+                                    unreachable!();
+                                }
+                                msg
                             }
                         }
+                    },
+                };
+
+                // We have a job to do, we must keep going until it's finished otherwise it will be lost with the rest of the channel
+                loop {
+                    match &mut connection {
+                        Some(connection) => {
+                            Client::handle_msg(msg, connection, &config).await;
+                            break;
+                        }
+                        None => connection = Client::attempt_connect(&config).await,
                     }
                 }
```

Comment:
I don't think this is going to exit properly if the shutdown message is sent. It looks like it will do the following:

1. The `shutdown_request_rx.recv()` future will fire.
2. Attempt to get a message off the `command_receiver` channel (which seems fine if it gets `TryRecvError::Empty` or any other error message as it breaks out of the `loop`).
3. Store the message from the channel as `msg`.
4. Attempt to send the message in the inner `loop`.
5. We'll then go back to the `let msg = tokio::select! {...}`.
6. Never execute the shutdown logic again as we've already pulled the message off the channel.

It's in the second time through the loop that there are issues. `shutdown_request_rx.recv()` should block forever, since we've already read the message that was in the channel, and nothing would (normally) be sending another one. This means that we're stuck in the loop waiting for `command_receiver.recv()`, and will never shut down the faktory connection task.

Reply:
Oh I just saw your message, I'm not sure I follow tho. Like 336 should solve that as it's re-filling the shutdown_request_rx.

Reply:
That's because I completely misread that line.